### ISSUE TITLE
Support for geopoint type

### ISSUE THREAD

**Issue by gunjpan** [strongloop/strong-remoting#365]
As noted in `strong-remoting@3.x` [release notes](https://github.com/strongloop/strong-remoting/releases/tag/v3.0.0), 

> Deep members of object arguments are no longer coerced. For example, a query like ?arg[count]=42 produces { count: '42' } now.

Thus,  `GeoPoint` datatype input in following `query` is treated as `string`:
`http://localhost:3010/api/gps/testGeoPoint?geopoint[lat]=23&geopoint[lng]=25`

  where as, following query works fine, (unusual:
`http://localhost:3010/api/gps/testGeoPoint?geopoint={"lat": 23, "lng":25}`

Also, there is no check for `GeoPoint` data type. What should be done in this case?
Possible solution: Add coercion check for `GeoPoint` data type

@bajtos ^ thoughts?

cc: @strongloop/loopback-dev

**Update 11/11/16**:
  - After discussing with @bajtos, we have decided to extract out `geo` class from [loopback-datasource-juggler/lib/geo.js](https://github.com/strongloop/loopback-datasource-juggler/blob/72c27b040c61e217d229b792e4db54275a034dec/lib/geo.js)  as an independent module and use it here for dealing with `GeoPoint` object.
See strongloop-internal/scrum-loopback#1176 for new module task

**Update 14/11/16**:
  - Everyone has a different way to create geopoint, @gunjpan will take geopoint class from `jugglar` and extract into its own module for use in `strong-remoting`.

**Update 15/11/16**:
  - Waiting on https://github.com/strongloop-internal/scrum-loopback/issues/1176 to land

**Update 16/11/16**:
  - PR that is pulled out into own module needs 2nd review to land, went back to https://github.com/strongloop/strong-remoting/pull/383 and requiring new module in it now.

**Update 17/11/16**:
  - PR landed for pulled out module: loopback-datatype-geopoint. Working on integrating this new module in https://github.com/strongloop/strong-remoting/pull/383. Some fundamental bugs with the new module, will fix and throw into own repo then publish maybe today/tomorrow.

**Update 18/11/16**
  - Will write docs by hand for #383, will try land today.

**Comment by bajtos** [comment]
@gunjpan could you please point me to the remoting metadata of your `gps/testGeoPoint` method?

> Thus,  `GeoPoint` datatype input in following `query` is treated as `string`:
`http://localhost:3010/api/gps/testGeoPoint?geopoint[lat]=23&geopoint[lng]=25`

LoopBack's GeoPoint type is automatically coercing string values to numbers in its constructor, see [lib/geo.js](https://github.com/strongloop/loopback-datasource-juggler/blob/99a864e9c6ef4907ed32425ef1f196d2677d17a8/lib/geo.js#L155-L158) in juggler. 

I think that as a result, the above query should work fine in most LoopBack applications.

> where as, following query works fine, (unusual:
`http://localhost:3010/api/gps/testGeoPoint?geopoint={"lat": 23, "lng":25}`

What do you find unusual?

> Add coercion check for GeoPoint data type

This makes a lot of sense to me. strong-remoting does not have any concept of `GeoPoint` type, it's defined in loopback-datasource-juggler. I think the correct solution is to modify `loopback` and register a type converter for `GeoPoint`. Note that this converter has to support both flavours of GeoPoint values: a string `23,25` and an object `{"lat":23,"lng":25}`.

Here is the place in LoopBack code where we create new `RemoteObjects` instance and where to register `GeoPoint` type: https://github.com/strongloop/loopback/blob/f9a58083ae247590c6f9541104200f6bc8d1b4cb/lib/application.js#L70

**Comment by gunjpan** [comment]
@bajtos :
> could you please point me to the remoting metadata of your gps/testGeoPoint method?

the method is same as defined [here](https://github.com/strongloop/loopback-sdk-ios/blob/master/LoopBackTests/server/index.js#L84-L93)

>What do you find unusual?

IMO, `http://localhost:3010/api/gps/testGeoPoint?geopoint[lat]=23&geopoint[lng]=25` is a more general way of querying.

**Comment by bajtos** [comment]
> could you please point me to the remoting metadata of your gps/testGeoPoint method?
the method is same as defined here

```js
Widget.testGeoPoint = function(geopoint, callback) {
  // add 1 to both lat and lng of the given geopoint and return it
  geopoint.lat += 1;
  geopoint.lng += 1;
  callback(null, geopoint);
};
Widget.testGeoPoint.shared = true;
Widget.testGeoPoint.accepts = [{ arg: 'geopoint', type: 'geopoint' }];
Widget.testGeoPoint.returns = [{ arg: 'geopoint', type: 'geopoint' }];
Widget.testGeoPoint.http = { verb: 'get' };
```

Thanks. Since `GeoPoint` is a valid LoopBack type, I think it's reasonable to expect that it can be used in remoting metadata too. Which means we should implement the change I proposed in my last comment.

> > What do you find unusual?

> IMO, `http://localhost:3010/api/gps/testGeoPoint?geopoint[lat]=23&geopoint[lng]=25` is a more general way of querying.

Both options are valid. The JSON encoding makes it easier to explicitly provide type information for deeply nested values.

**Comment by gunjpan** [comment]
Landed #383.
